### DISCUSSION TITLE
How to download temperature_2m for GFS 0.5 Ensemble using Docker?

### DISCUSSION THREAD

**Discussion by magheralex** [open-meteo/open-meteo#684] [Q&A]
Hi there,
When downloading the temperature_2m variable for the GFS 0.5 degrees Ensemble model, it's expected to retrieve data for the entire forecast period, typically spanning 35 days.

![image](https://github.com/open-meteo/open-meteo/assets/30143275/e243f858-11d2-442d-aabc-8cc0497faed3)

Running on Docker:
`docker run -it --rm -v open-meteo-data:/app/data ghcr.io/open-meteo/open-meteo download-gfs gfs05_ens --only-variables temperature_2m`

Is this the correct command?

However, I'm only receiving data for 16 days, instead of 35. Moreover, the first 13 values are null:

```
[ INFO ] Finished downloading 765.0 MB in 02:00
[ INFO ] Convert timerange 2024-03-07T12:00 to 2024-03-23T12:00 (3-hourly)
[ INFO ] [ Convert temperature_2m ] Starting
[ INFO ] [ Convert temperature_2m ] 20% 1660608 / 8057520 in 22s, 76162/s remaining 84s
[ INFO ] [ Convert temperature_2m ] 38% 3113640 / 8057520 in 42s, 70981/s remaining 70s
[ INFO ] [ Convert temperature_2m ] 59% 4774248 / 8057520 in 64s, 76830/s remaining 43s
[ INFO ] [ Convert temperature_2m ] 79% 6434856 / 8057520 in 84s, 81376/s remaining 19.9s
[ INFO ] [ Convert temperature_2m ] 100% 8057520 / 8057520 in 106s, 74370/s remaining 0.0ms
[ INFO ] [ Convert temperature_2m ] Completed in 108s
[ INFO ] Conversion completed in 109s
```

When I access the url /v1/ensemble?latitude=y&longitude=x&hourly=temperature_2m&timezone=auto&forecast_days=30&models=gfs05, the API returns:

`"temperature_2m":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,9.5,9.6,9.6,9.4,8.9,8.3,7.7,7.2,6.9,6.5,6.1,5.6,5.2,4.9,4.7,4.4,4.2,4.1,4.2,4.9,6.1,7.1,7.9,8.6,9.2,9.4,9.5,9.3,8.7,7.9,7.1,6.4,5.8,5.3,4.9,4.5,4.2,3.8,3.5,3.3,3.3,3.6,4.2,5.6,7.4,8.9,10.0,10.8,11.4,11.9,12.1,11.9,11.3,10.4,9.5,8.9,8.4,7.8,7.2,6.7,6.2,5.6,5.1,4.8,4.8,4.9,5.6,7.2,9.2,11.1,12.6,13.9,14.9,15.5,15.9,15.6,14.6,13.1,11.6,10.6,9.6,8.8,8.1,7.5,7.1,6.7,6.4,6.4,6.4,6.7,7.2,8.4,10.0,11.6,13.1,14.6,15.9,17.0,17.9,18.1,17.2,15.6,14.4,13.9,13.6,13.5,13.3,13.1,13.0,12.8,12.6,12.4,12.4,12.4,12.6,13.2,14.1,15.1,16.1,17.2,17.9,17.9,17.6,17.0,16.2,15.2,14.1,13.0,11.9,11.0,10.6,10.5,10.4,10.3,10.1,10.1,10.1,10.2,10.3,10.4,10.4,10.5,10.5,10.5,10.7,11.2,12.0,12.4,12.1,11.4,10.8,10.2,9.6,9.1,8.8,8.6,8.4,8.0,7.6,7.3,7.4,7.7,8.4,10.1,12.2,14.1,15.6,16.9,17.7,18.0,18.0,17.6,16.6,15.1,13.9,13.1,12.5,12.0,11.6,11.3,11.0,10.5,9.9,9.6,9.6,9.7,10.2,11.4,12.9,14.3,15.6,16.8,17.7,18.1,18.3,18.0,17.2,16.0,15.1,14.5,14.1,13.8,13.3,12.8,12.4,11.9,11.4,11.2,11.3,11.6,12.2,13.6,15.2,16.6,17.6,18.4,18.8,19.0,18.9,18.6,18.2,17.7,17.2,16.9,16.5,16.1,15.4,14.8,14.1,13.4,12.8,12.4,12.2,12.4,12.9,13.9,15.5,16.9,18.0,18.9,19.5,19.5,19.2,18.8,18.1,17.5,16.8,16.3,15.9,15.4,15.0,14.6,14.2,13.8,13.3,12.9,12.7,12.6,12.9,13.7,14.9,16.1,17.4,18.5,19.2,19.2,18.9,18.4,17.7,16.9,16.1,15.3,14.5,13.7,12.9,12.2,11.6,11.0,10.5,10.2,10.0,10.0,10.4,11.6,13.3,14.9,16.5,18.1,19.0,19.0,18.4,17.6,16.6,15.6,14.4,13.5,12.6,11.7,10.9,10.1,9.5,9.0,8.6,8.4,8.4,8.6,9.1,10.3,11.9,13.5,15.1,16.5,17.4,17.4,16.7,15.9,14.9,13.9,12.8,11.9,11.1,10.4,9.8,9.1,8.6,8.3,8.1,7.9,7.9,8.1,8.4,9.2,10.2,11.2,12.1,13.1,13.6,13.6,13.2,12.8,12.2,11.6,11.1,10.7,10.4,10.1,9.9,9.7,9.5,9.4,9.2,9.1,9.1,9.0,9.0,9.0,9.0,9.1,9.1,9.1,9.2,9.3,9.4,9.6,9.7,9.8,9.8,9.6,9.4,9.1,8.9,8.6,8.4,8.4,8.2,8.2,8.2,8.3,8.4,8.4,8.6,8.7,8.9,9.0,9.1,9.1,9.1,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null]`

During downloading the data, I got several "download failed" errors, even if the files are hosted on the nomads.ncep.noaa.gov server:
![image](https://github.com/open-meteo/open-meteo/assets/30143275/d988af22-941d-4dd7-9cff-ab18f3dcafd0)

So, why does the GFS 0.5 degrees Ensemble model data for temperature_2m only provide 16 days instead of the full 35-day forecast?

Thanks in advance for your help.

**Comment by patrick-zippenfenig** [comment]
Hi,

the GFS ensemble downloader uses multiple steps to download the entire time range. You have to call it with `--second-flush` to download days 16-35. This is due to different update timings. Here is the required cronjob:

```55 23 * * * /usr/local/bin/openmeteo-api download-gfs gfs05_ens --second-flush --run 0 > ~/log/gfs05_ens-second-flush.log 2>&1 || cat ~/log/gfs05_ens-second-flush.log```

**Comment by magheralex** [comment]
Hi Patrick,
Thanks for your answer.

I've tried to download it with 
`docker run -it --rm -v open-meteo-data:/app/data ghcr.io/open-meteo/open-meteo download-gfs gfs05_ens --only-variables temperature_2m --second-flush`

and got 404 for the 1st hour from day 16 (forecastHour 390):
![image](https://github.com/open-meteo/open-meteo/assets/30143275/49c965c1-d9a1-41da-940c-8b39f90578eb)

I can't download any data starting from hour 390.

**Reply by patrick-zippenfenig** [reply]
Please also set `--run 0`. Otherwise the system tries to download 12z, which does not offer the extended period

**Reply by magheralex** [reply]
`docker run -it --rm -v open-meteo-data:/app/data ghcr.io/open-meteo/open-meteo download-gfs gfs05_ens --only-variables temperature_2m --second-flush --run 0` returns the same 404:
![image](https://github.com/open-meteo/open-meteo/assets/30143275/5ec5889b-ad04-43ef-9c6d-16627661a97c)

**Reply by patrick-zippenfenig** [reply]
It looks like the 0z control run has not been publish on the NCEP server yet, while individual members have been published. This is a bit strange, but otherwise correct. The downloader expects delays and will now try to download the required files every 2 minutes. Once the required files are published in the NCEP server, processing starts

**Reply by magheralex** [reply]
Hi Patrick,
I've tried again. They published the data for 11.03.2024 on the 10.03.2024 folder: https://nomads.ncep.noaa.gov/pub/data/nccf/com/gens/prod/gefs.20240310/00/atmos/pgrb2ap5/

![image](https://github.com/open-meteo/open-meteo/assets/30143275/5f174c04-5abf-411f-bd9d-e57679fd49f2)

However, the downloader is trying to search within gefs.20240311 (where is no 0z control run) instead of gefs.20240310 folder.

There's the same pattern for the previous day. The 0z control run for 10 March is on the 09.03 folder:
![image](https://github.com/open-meteo/open-meteo/assets/30143275/f383d960-a3ff-4528-b79e-e5cd058e9fb9)

Maybe that's why I get 404.

**Reply by magheralex** [reply]
Initially, I thought NOAA had made a mistake but I noticed a daily pattern of publishing 0z control run in the previous day's folder. 

![image](https://github.com/open-meteo/open-meteo/assets/30143275/6f46629b-1499-4ea9-a9d3-f6c1903dde1c)

How can I download 0z control run data from the previous day folder?

**Reply by patrick-zippenfenig** [reply]
`--run 2024031100` selects an individual run with an absolute date+run timestamp

**Reply by magheralex** [reply]
Thanks, Patrick,
It works.

I'm trying to download the 00-run for ECMWF IFS025 as well, but it seems the downloader uses an invalid path.
![image](https://github.com/open-meteo/open-meteo/assets/30143275/02eb77f6-d099-44f2-bb40-7bfd3cdd3606)

Cmd used: `docker run -it --rm -v open-meteo-data:/app/data ghcr.io/open-meteo/open-meteo download-ecmwf --domain ifs025_ensemble --only-variables temperature_2m --run 0`

The correct path is https://data.ecmwf.int/forecasts/20240322/00z/ifs/0p25/enfo/

/ifs/ is missing from the current downloader (used for downloading the 00-run): https://data.ecmwf.int/forecasts/20240322/00z/0p25/enfo/

Can you help me with this? Thanks in advance!

**Reply by patrick-zippenfenig** [reply]
ECMWF changed URL paths for their open-data servers a couple of weeks back. Please make sure to use the latest [Open-Meteo Docker image](https://github.com/open-meteo/open-meteo/pkgs/container/open-meteo). Such changes are rapidly applied to Open-Meteo, but users of the Docker images have to adapt to changes themself.

**Reply by magheralex** [reply]
Thanks @patrick-zippenfenig 
It works.

Can I also download the weather_code for ECMWF ifs025_ensemble and gfs05_ens?

It seems there is no weather_code variable for ECMWF IFS025 Ensemble:

Cmd:
`docker run -it --rm -v open-meteo-data:/app/data ghcr.io/open-meteo/open-meteo download-ecmwf --domain ifs025_ensemble --only-variables temperature_2m,weather_code --run 0
`
Error:

[ WARNING ] App.RawRepresentableError.invalidValue(value: "weather_code", availa                                   bleValues: ["precipitation", "runoff", "soil_temperature_0_to_7cm", "surface_tem                                   perature", "geopotential_height_1000hPa", "geopotential_height_925hPa", "geopote                                   ntial_height_850hPa", "geopotential_height_700hPa", "geopotential_height_500hPa"                                   , "geopotential_height_300hPa", "geopotential_height_250hPa", "geopotential_heig                                   ht_200hPa", "geopotential_height_50hPa", "wind_v_component_1000hPa", "wind_v_com                                   ponent_925hPa", "wind_v_component_850hPa", "wind_v_component_700hPa", "wind_v_co                                   mponent_500hPa", "wind_v_component_300hPa", "wind_v_component_250hPa", "wind_v_c                                   omponent_200hPa", "wind_v_component_50hPa", "wind_u_component_1000hPa", "wind_u_                                   component_925hPa", "wind_u_component_850hPa", "wind_u_component_700hPa", "wind_u                                   _component_500hPa", "wind_u_component_300hPa", "wind_u_component_250hPa", "wind_                                   u_component_200hPa", "wind_u_component_50hPa", "temperature_1000hPa", "temperatu                                   re_925hPa", "temperature_850hPa", "temperature_700hPa", "temperature_500hPa", "t                                   emperature_300hPa", "temperature_250hPa", "temperature_200hPa", "temperature_50h                                   Pa", "relative_humidity_1000hPa", "relative_humidity_925hPa", "relative_humidity                                   _850hPa", "relative_humidity_700hPa", "relative_humidity_500hPa", "relative_humi                                   dity_300hPa", "relative_humidity_250hPa", "relative_humidity_200hPa", "relative_                                   humidity_50hPa", "surface_pressure", "pressure_msl", "total_column_integrated_wa                                   ter_vapour", "wind_v_component_10m", "wind_u_component_10m", "specific_humidity_                                   1000hPa", "specific_humidity_925hPa", "specific_humidity_850hPa", "specific_humi                                   dity_700hPa", "specific_humidity_500hPa", "specific_humidity_300hPa", "specific_                                   humidity_250hPa", "specific_humidity_200hPa", "specific_humidity_50hPa", "temper                                   ature_2m", "relative_vorticity_1000hPa", "relative_vorticity_925hPa", "relative_                                   vorticity_850hPa", "relative_vorticity_700hPa", "relative_vorticity_500hPa", "re                                   lative_vorticity_300hPa", "relative_vorticity_250hPa", "relative_vorticity_200hP                                   a", "relative_vorticity_50hPa", "divergence_of_wind_1000hPa", "divergence_of_win                                   d_925hPa", "divergence_of_wind_850hPa", "divergence_of_wind_700hPa", "divergence                                   _of_wind_500hPa", "divergence_of_wind_300hPa", "divergence_of_wind_250hPa", "div                                   ergence_of_wind_200hPa", "divergence_of_wind_50hPa", "cloud_cover", "cloud_cover                                   _low", "cloud_cover_mid", "cloud_cover_high"])
Swift/ErrorType.swift:200: Fatal error: Error raised at top level: App.RawRepres                                   entableError.invalidValue(value: "weather_code", availableValues: ["precipitatio                                   n", "runoff", "soil_temperature_0_to_7cm", "surface_temperature", "geopotential_                                   height_1000hPa", "geopotential_height_925hPa", "geopotential_height_850hPa", "ge                                   opotential_height_700hPa", "geopotential_height_500hPa", "geopotential_height_30                                   0hPa", "geopotential_height_250hPa", "geopotential_height_200hPa", "geopotential                                   _height_50hPa", "wind_v_component_1000hPa", "wind_v_component_925hPa", "wind_v_c                                   omponent_850hPa", "wind_v_component_700hPa", "wind_v_component_500hPa", "wind_v_                                   component_300hPa", "wind_v_component_250hPa", "wind_v_component_200hPa", "wind_v                                   _component_50hPa", "wind_u_component_1000hPa", "wind_u_component_925hPa", "wind_                                   u_component_850hPa", "wind_u_component_700hPa", "wind_u_component_500hPa", "wind                                   _u_component_300hPa", "wind_u_component_250hPa", "wind_u_component_200hPa", "win                                   d_u_component_50hPa", "temperature_1000hPa", "temperature_925hPa", "temperature_                                   850hPa", "temperature_700hPa", "temperature_500hPa", "temperature_300hPa", "temp                                   erature_250hPa", "temperature_200hPa", "temperature_50hPa", "relative_humidity_1                                   000hPa", "relative_humidity_925hPa", "relative_humidity_850hPa", "relative_humid                                   ity_700hPa", "relative_humidity_500hPa", "relative_humidity_300hPa", "relative_h                                   umidity_250hPa", "relative_humidity_200hPa", "relative_humidity_50hPa", "surface                                   _pressure", "pressure_msl", "total_column_integrated_water_vapour", "wind_v_comp                                   onent_10m", "wind_u_component_10m", "specific_humidity_1000hPa", "specific_humid                                   ity_925hPa", "specific_humidity_850hPa", "specific_humidity_700hPa", "specific_h                                   umidity_500hPa", "specific_humidity_300hPa", "specific_humidity_250hPa", "specif                                   ic_humidity_200hPa", "specific_humidity_50hPa", "temperature_2m", "relative_vort                                   icity_1000hPa", "relative_vorticity_925hPa", "relative_vorticity_850hPa", "relat                                   ive_vorticity_700hPa", "relative_vorticity_500hPa", "relative_vorticity_300hPa",                                    "relative_vorticity_250hPa", "relative_vorticity_200hPa", "relative_vorticity_5                                   0hPa", "divergence_of_wind_1000hPa", "divergence_of_wind_925hPa", "divergence_of                                   _wind_850hPa", "divergence_of_wind_700hPa", "divergence_of_wind_500hPa", "diverg                                   ence_of_wind_300hPa", "divergence_of_wind_250hPa", "divergence_of_wind_200hPa",                                    "divergence_of_wind_50hPa", "cloud_cover", "cloud_cover_low", "cloud_cover_mid",                                    "cloud_cover_high"])
